collect_app/src/main/java/org/odk/collect/android/widgets/AudioController.java

```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright (C) 2018 Shobhit Agarwal
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.odk.collect.android.widgets;
+
+import android.content.Context;
+import android.media.AudioManager;
+import android.media.MediaPlayer;
+import android.net.Uri;
+import android.os.Handler;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.widget.ImageButton;
+import android.widget.SeekBar;
+import android.widget.TextView;
+
+import org.javarosa.form.api.FormEntryPrompt;
+import org.odk.collect.android.R;
+import org.odk.collect.android.activities.FormEntryActivity;
+import org.odk.collect.android.application.Collect;
+import org.odk.collect.android.utilities.DateTimeUtils;
+
+import java.io.File;
+import java.io.IOException;
+
+import butterknife.BindView;
+import butterknife.ButterKnife;
+import butterknife.OnClick;
+import timber.log.Timber;
+
+import static android.view.View.GONE;
+
+public class AudioController implements SeekBar.OnSeekBarChangeListener {
+
+    private static final int SEEK_FORWARD_TIME = 5000; // 5 seconds
+    private static final int SEEK_BACKWARD_TIME = 5000; // 5 seconds
+
+    @BindView(R.id.currentDuration)
+    TextView currentDurationLabel;
+    @BindView(R.id.totalDuration)
+    TextView totalDurationLabel;
+    @BindView(R.id.playBtn)
+    ImageButton playButton;
+    @BindView(R.id.seekBar)
+    SeekBar seekBar;
+
+    private Context context;
+    private MediaPlayer mediaPlayer;
+    private Handler seekHandler = new Handler();
+
+    /**
+     * Background Runnable thread
+     */
+    private Runnable updateTimeTask = new Runnable() {
+        public void run() {
+            try {
+                if (mediaPlayer.isPlaying()) {
+                    updateTimer();
+                }
+                seekHandler.postDelayed(this, 100);
+            } catch (IllegalStateException e) {
+                seekHandler.removeCallbacks(updateTimeTask);
+                Timber.i(e, "Attempting to update timer when player is stopped");
+            }
+        }
+    };
+
+    private State state;
+    private FormEntryPrompt formEntryPrompt;
+    private View view;
+
+    AudioController() {
+    }
+
+    void init(Context context, MediaPlayer mediaPlayer, FormEntryPrompt formEntryPrompt) {
```

Comment:
I wish the fields could be final and set in the constructor. Do you need this `init` for testing?

Reply:
Only context can be set in the constructor as `mediaPlayer` and `formEntryPrompt` can only be set once the super constructor has been called

Reply:
OK. Thanks.